---
title: "SNP FILTERING"
output:
  html_notebook:
    code_folding: hide
    df_print: paged
    highlight: kate
    theme: yeti
    toc: no
  html_document:
    toc: no
---

To use this script, find and replace 20181125 with the current project date

```{r load libraries}

# knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(readr)
library(ggplot2)
library(dplyr)
# source("~/02-apcl-ddocent/APCL_analysis/scr/libraries.R")
source("~/02-apcl-ddocent/APCL_analysis/scr/ggplot.R")
source("~/02-apcl-ddocent/APCL_analysis/scr/VCFfilterstats.R")
source("~/02-apcl-ddocent/APCL_analysis/scr/xtrafunctions.R")
```

 
```{bash}
mkdir /local/home/michelles/02-apcl-ddocent/APCL_analysis/20181125/results
```


## Raw data stats

### Query stats

Query raw stats using vcftools.

```{bash}
cd /local/home/michelles/02-apcl-ddocent/APCL_analysis/20181125
# depth indv/locus
vcftools --vcf /data/apcl/all_samples/20181125/TotalRawSNPs.vcf --out results/raw --depth
vcftools --vcf /data/apcl/all_samples/20181125/TotalRawSNPs.vcf  --out results/raw --site-mean-depth

# missing data indv/locus
vcftools --vcf /data/apcl/all_samples/20181125/TotalRawSNPs.vcf  --out results/raw --missing-indv
vcftools --vcf /data/apcl/all_samples/20181125/TotalRawSNPs.vcf  --out results/raw --missing-site

# heterozygosity per individual
vcftools --vcf /data/apcl/all_samples/20181125/TotalRawSNPs.vcf --out results/raw --het

# SNP call quality
vcftools --vcf /data/apcl/all_samples/20181125/TotalRawSNPs.vcf --out results/raw --site-quality
```

### Visualize

```{r fig.height=20, fig.width=10}

# load stats files ----
ind_stats_raw <- read.ind.stats(dir = "results", vcf = "raw")
loc_stats_raw <- read.loc.stats(dir = "results", vcf = "raw")

# plot missing data per indv ----
p1 <- ggplot(ind_stats_raw, aes(x = F_MISS)) +
  geom_histogram(binwidth = .01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "missing data per indv") +
  theme_standard

# plot Fis per indv ----
# p2 <- ggplot(ind_stats_raw, aes(x = F)) +
#   geom_histogram(binwidth = .01, color = "black", fill = "grey95") +
#   geom_vline(aes(xintercept = mean(F, na.rm = TRUE)),
#                  color = "red", linetype = "dashed", size = 1) +
#   geom_vline(aes(xintercept = 0),
#                  color = "darkblue", linetype = "dashed", size = 1) +
#   labs(x = "Fis per indv") +
#   theme_standard

# plot read depth per indv ----
p3 <- ggplot(ind_stats_raw, aes(x = MEAN_DEPTH)) +
  geom_histogram(binwidth = 10, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean read depth per indv") +
  theme_standard

# plot depth vs missing ----
p4 <- ggplot(ind_stats_raw, aes(x = MEAN_DEPTH, y = F_MISS)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per indv", y = "% missing data") +
  theme_standard

# plot Fis vs missing data per indv ----
# p5 <- ggplot(ind_stats_raw, aes(x = F, y = F_MISS)) +
#   geom_point() +
#   geom_vline(aes(xintercept = mean(F, na.rm = TRUE)),
#                  color = "red", linetype = "dashed", size = 1) +
#   geom_vline(aes(xintercept = 0),
#                  color = "darkblue", linetype = "dashed", size = 1) +
#   geom_hline(aes(yintercept = mean(F_MISS, na.rm = TRUE)),
#                  color = "red", linetype = "dashed", size = 1) +
#   geom_hline(aes(yintercept = 0.5),
#                  color = "darkblue", linetype = "dashed", size = 1) +
#   labs(x = "Fis per indv", y = "% missing data") +
#   theme_standard

# plot Fis vs mean depth per indv ----
# p6 <- ggplot(ind_stats_raw, aes(x = F, y = MEAN_DEPTH)) +
#   geom_point() +
#   geom_vline(aes(xintercept = mean(F, na.rm = TRUE)),
#                  color = "red", linetype = "dashed", size = 1) +
#   geom_vline(aes(xintercept = 0),
#                  color = "darkblue", linetype = "dashed", size = 1) +
#   geom_hline(aes(yintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
#                  color = "red", linetype = "dashed", size = 1) +
#   geom_hline(aes(yintercept = 20),
#                  color = "darkblue", linetype = "dashed", size = 1) +
#   labs(x = "Fis per indv", y = "mean depth per indv") +
#   theme_standard

# plot distribution missing data per locus ----
p7 <- ggplot(loc_stats_raw, aes(x = F_MISS)) +
  geom_histogram(binwidth = 0.01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0.1),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "% missing data per locus") +
  theme_standard

# plot distribution mean read depth ----
p8 <- ggplot(loc_stats_raw, aes(x = MEAN_DEPTH)) +
  geom_histogram(binwidth = 5, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean read depth per locus") +
  theme_standard

# plot read depth vs missing data ----
p9 <- ggplot(loc_stats_raw, aes(x = MEAN_DEPTH, y = F_MISS)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.1),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per locus", y = "% missing data") +
  theme_standard

# plot no of SNPs per locus ----
p10 <- loc_stats_raw %>%
  count(CHROM) %>%
  ggplot(aes(x = n)) +
  geom_histogram(binwidth = 1, color = "black", fill = "grey95") + 
  labs(x = "number of SNPs per locus") +
  theme_standard

temp <- loc_stats_raw %>%
  count(CHROM)

# plot number of SNPs per contig vs. mean depth ----
p11 <- left_join(temp, loc_stats_raw) %>%
  ggplot() +
  geom_point(aes(x = n, y = MEAN_DEPTH)) +
  labs(x = "number of SNPs per contig", y = "mean depth") +
  theme_standard

# plot depth vs SNP quality ----
site_qual <- read.table("results/raw.lqual", 
                        header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(PROB = 10^(-QUAL/10))

temp <- data.frame(loc_stats_raw$MEAN_DEPTH, site_qual$QUAL) %>%
  rename(depth = loc_stats_raw.MEAN_DEPTH, qual = site_qual.QUAL)

p12 <- ggplot(temp, aes(x = depth, y = qual)) +
  geom_point(size = 1) +
  geom_vline(aes(xintercept = mean(depth, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(qual, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per locus", y = "SNP quality") +
  theme_standard

m1 <- multiplot(p1, p3, p4, p7, p8, p9, p10, p11, p12, cols=2)
# m1 <- multiplot(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, p11, p12, cols=2)

```

Data set contains `r nrow(ind_stats_raw)` individuals and `r nrow(loc_stats_raw)` loci.


## Run SNP counting script

```{bash, eval=FALSE, include=FALSE}

echo "FILTER SNP CONTIG INDV" > Filter.count

  SNP=$(grep -cv '#' results/raw.recode.vcf)
  CONTIG=$(grep -v '#' results/raw.recode.vcf | cut -f 1 | sort | uniq | wc -l)
  INDV=$(vcfsamplenames results/raw.recode.vcf | wc -l)
  echo "raw.recode.vcf $SNP $CONTIG $INDV" >> Filter.count


```

Compare SNPs/contigs/indv at each filtering step and between filtering schemes.

```{r}

count <- read.table("/local/home/michelles/02-apcl-ddocent/APCL_analysis/20181125/Filter.count", header = TRUE, stringsAsFactors = FALSE)
```