---
title: "SNP FILTERING"
output:
  github_document: default
  html_notebook:
    code_folding: hide
    df_print: paged
    highlight: kate
    theme: yeti
    toc: no
  html_document:
    toc: no
---
This filtering scheme follows the FS6 scheme of SJO'Leary but stops the alternate paring of genotypes and missing individuals with missing data at the 70% / 70% step.


```{r setup, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
	echo = FALSE,
	message = FALSE,
	warning = FALSE
)
library(ggplot2)
library(readr)
library(dplyr)
library(here)
source(here("filtering", "ggplot.R"))
source(here("filtering", "VCFfilterstats.R"))
source(here("filtering", "xtrafunctions.R"))


```

## Filtering to yield more loci than FS6

### Step 1: Filter LQ loci

Remove SNPs with quality score < 20, minimum depth per genotype < 5, minimum mean depth < 15.  
This chunk can take some time because it is filtering out the full data set. Adding maf 0.2 here results in ~1000 SNPs, trying to add it later.

```{bash include=FALSE}
vcftools --vcf /data/apcl/all_samples/20181127/TotalRawSNPs.vcf --out results/step1 --minQ 20 --minDP 5 --min-meanDP 15 --recode --recode-INFO-all
```

```{bash echo=FALSE}
echo "FILTER SNP CONTIG INDV" > Filter.count
  FILTER=$(echo "step1")
  SNP=$(grep -cv '#' results/step1.recode.vcf)
  CONTIG=$(grep -v '#' results/step1.recode.vcf | cut -f 1 | sort | uniq | wc -l)
  INDV=$(vcfsamplenames results/step1.recode.vcf | wc -l)
  echo "step1.recode.vcf $FILTER $SNP $CONTIG $INDV" >> Filter.count


```

### Step 2: Missing data indvidiuals/loci 
The max-missing option excludes sites based on the proportion of missing data where 0 is completely missing and 1 is not missing at all.

#### 2a: Retain loci with genotype call rate > 50% and flag individuals with missing data.
```{bash include=FALSE}

vcftools --vcf results/step1.recode.vcf --out results/step2a --max-missing 0.5 --recode --recode-INFO-all

vcftools --vcf results/step2a.recode.vcf --out results/it1 --missing-indv

```

```{bash echo=FALSE}
  FILTER=$(echo "step2a")
  SNP=$(grep -cv '#' results/step2a.recode.vcf)
  CONTIG=$(grep -v '#' results/step2a.recode.vcf | cut -f 1 | sort | uniq | wc -l)
  INDV=$(vcfsamplenames results/step2a.recode.vcf | wc -l)
  echo "step2a.recode.vcf $FILTER $SNP $CONTIG $INDV" >> Filter.count
```


#### 2b: Identify individuals with > 90% missing data.  
Everything to the right of the blue line will be eliminated.

```{r message=FALSE}
imiss <- read_delim(here("filtering","results", "it1.imiss"), delim = "\t")

ggplot(imiss, aes(x = F_MISS)) +
  geom_histogram(binwidth = 0.05, color = "black", fill = "grey75") +
  geom_vline(xintercept = 0.9, color = "darkblue", linetype = "dashed") +
  scale_x_continuous(limits = c(0, 1)) +
  theme_standard

LQ_indv <- imiss %>%
  filter(F_MISS > 0.9) %>%
  select(INDV)

write.table(LQ_indv, here("filtering","results", "LQ_Ind_it1a"),
            col.names = FALSE, row.names = FALSE, quote = FALSE)

```

Remove flagged individuals, adding maf 0.2 here results in 0 SNPs, trying later.

```{bash include=FALSE}

vcftools --vcf results/step2a.recode.vcf --out results/step2b --remove results/LQ_Ind_it1a --recode --recode-INFO-all
```


```{bash echo=FALSE}
  FILTER=$(echo "step2b")
  SNP=$(grep -cv '#' results/step2b.recode.vcf)
  CONTIG=$(grep -v '#' results/step2b.recode.vcf | cut -f 1 | sort | uniq | wc -l)
  INDV=$(vcfsamplenames results/step2b.recode.vcf | wc -l)
  echo "step2b.recode.vcf $FILTER $SNP $CONTIG $INDV" >> Filter.count
```

### Step 3: Info flag filtering

#### Allele balance

AB: Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous

```{bash include=FALSE}

# allele balance
cut -f8 results/step2b.recode.vcf | grep -oe "AB=[[:digit:]].[[:digit:]][[:digit:]][[:digit:]]" | sed -s 's/AB=//g' > results/step3.AB

# site Depth
cut -f8 results/step2b.recode.vcf | grep -oe "DP=[0-9]*" | sed -s 's/DP=//g' > results/step3.DEPTH

# quality score
mawk '!/#/' results/step2b.recode.vcf | cut -f1,2,6 > results/step3.QUAL

# mapping quality
cut -f8 results/step2b.recode.vcf | grep -oe "MQM=[0-9]*" | sed -s 's/MQM=//g' > results/step3.MQM
cut -f8 results/step2b.recode.vcf | grep -oe "MQMR=[0-9]*" | sed -s 's/MQMR=//g' > results/step3.MQMR

```

Allele balance is the ratio of reads for reference allele to all reads, considering only reads from individuals called as heterozygous. Values range from 0 - 1; allele balance (for real loci) should be approx. 0.5. Filter contigs SNPs for which the with allele balance < 0.25 and > 0.75.

```{r}
read.table(here("filtering","results", "step3.AB"),
           col.names = "AB", stringsAsFactors = FALSE) %>%
  ggplot(aes(x = AB)) +
  geom_histogram(binwidth = 0.01, color = "black", fill = "grey95") +
  geom_vline(xintercept = 0.5, color = "red", linetype = "dashed") +
  geom_vline(xintercept = 0.2, color = "darkblue", linetype = "dashed") +
  geom_vline(xintercept = 0.8, color = "darkblue", linetype = "dashed") +
  labs(x = "Allele balance ") +
  theme_standard
```

Filter contigs with SNP calls with AB > 0.2, AB > 0.8; retain loci very close to 0 (retain loci that are fixed variants). Remove genotypes if the quality sum of the reference or alternate allele was 0.

```{bash include=FALSE}
vcffilter -s -f "AB > 0.2 & AB < 0.8 | AB < 0.01 | AB > 0.99" -s -g "QR > 0 | QA > 0"  results/step2b.recode.vcf >  results/step3AB.recode.vcf 

mawk '!/#/' results/step3AB.recode.vcf | wc -l

echo $?

# allele balance
cut -f8 results/step3AB.recode.vcf | grep -oe "AB=[[:digit:]].[[:digit:]][[:digit:]][[:digit:]]" | sed -s 's/AB=//g' > results/step3AB.AB

```

#### replot to make sure the changes were effective
```{r message=FALSE}
read.table(here("filtering","results", "step3AB.AB"),
           col.names = "AB", stringsAsFactors = FALSE) %>%
  ggplot(aes(x = AB)) +
  geom_histogram(binwidth = 0.01, color = "black", fill = "grey95") +
  geom_vline(xintercept = 0.5, color = "red", linetype = "dashed") +
  geom_vline(xintercept = 0.2, color = "darkblue", linetype = "dashed") +
  geom_vline(xintercept = 0.8, color = "darkblue", linetype = "dashed") +
  labs(x = "Allele balance ") +
  theme_standard

```

#### Quality/depth ratio

Compare quality/depth.

```{r, fig.height=10, fig.width=8, message=FALSE, warning=FALSE}
# depth
depth <- read.table(here("filtering","results", "step3.DEPTH"),
                    col.names = "depth")

# quality score
qual <- read.table(here("filtering","results", "step3.QUAL"),
                   col.names = c("locus", "pos", "qual"))

df <- bind_cols(qual, depth) %>%
  mutate(ratio = qual/depth)

ggplot(df, aes(x = ratio)) + 
  geom_histogram(binwidth = 0.1, color = "black", fill = "grey95") +
  geom_vline(xintercept = 0.2, color = "darkred", linetype = "dashed") +
  geom_vline(xintercept = 0.5, color = "darkred", linetype = "dashed") +
  labs(x = "ratio depth/quality ", y = "no. of loci") +
  theme_standard

```

Remove loci with quality/depth ratio < 0.2

```{bash include=FALSE}
vcffilter -s -f "QUAL / DP > 0.2" results/step3AB.recode.vcf > results/step3ABQDPTH.recode.vcf

mawk '!/#/' results/step3ABQDPTH.recode.vcf | wc -l

echo $?

# site Depth
cut -f8 results/step3ABQDPTH.recode.vcf | grep -oe "DP=[0-9]*" | sed -s 's/DP=//g' > results/step3ABQDPTH.DEPTH

# quality score
mawk '!/#/' results/step3ABQDPTH.recode.vcf | cut -f1,2,6 > results/step3ABQDPTH.QUAL


```

```{r fig.height=10, fig.width=8}
# depth
depth <- read.table(here("filtering","results", "step3ABQDPTH.DEPTH"),
                    col.names = "depth")

# quality score
qual <- read.table(here("filtering","results", "step3ABQDPTH.QUAL"),
                   col.names = c("locus", "pos", "qual"))

df <- bind_cols(qual, depth) %>%
  mutate(ratio = qual/depth)

ggplot(df, aes(x = ratio)) + 
  geom_histogram(binwidth = 0.1, color = "black", fill = "grey95") +
  geom_vline(xintercept = 0.2, color = "darkred", linetype = "dashed") +
  geom_vline(xintercept = 0.5, color = "darkred", linetype = "dashed") +
  labs(x = "ratio depth/quality ", y = "no. of loci") +
  theme_standard

```
Based on the above graph, I was going to re-do the cutoff at 0.5 but the y-axis shows that the peak went down from 3000 to 150 so even though there is still a peak, it is clearly very zoomed in.


#### ratio mapping quality

Remove loci based on ratio of mapping quality for reference and alternate allele, i.e. sites that have a high discrepancy between the mapping qualities of two alleles.  Skipping this step because vcf tools takes more than the red circles.

```{r}

temp <- read.table(here("filtering","results", "step3.MQM"), col.names = "MQM")

mapqual <- read.table(here("filtering","results", "step3.MQMR"), col.names = "MQMR")

mapqual <- bind_cols(mapqual, temp) %>%
  mutate(ratio = MQM/MQMR)

filter <- mapqual %>%
  filter(ratio < 0.25 | ratio > 1.75)

ggplot(mapqual, aes(x = MQM, y = MQMR)) +
  geom_point(shape = 1) + 
  geom_abline(intercept = 0, slope = 1, size = 1, color = "red", linetype = "dashed") +
  geom_abline(intercept = 0, slope = 4, size = 1, color = "darkblue", linetype = "dashed") +
  geom_abline(intercept = 0, slope = 0.571, size = 1, color = "darkblue", linetype = "dashed") +
  geom_point(data = filter, aes(x = MQM, y = MQMR), shape = 1, color = "red") +
  scale_x_continuous(limits = c(0, 65)) + 
  scale_y_continuous(limits = c(0, 65)) +
  labs(x = "mean mapping quality alt allele ", y = "mean mapping quality ref allele ") +
  theme_standard

```

Filter loci with mapping quality ratio < 0.25 and > 1.75.

<!-- ```{bash include=FALSE} -->
<!--  -->
<!-- vcffilter -s -f "MQM / MQMR > 0.25 & MQM / MQMR < 1.75" results/step3ABQDPTH.recode.vcf > results/step3ABQDPTHMQM.recode.vcf -->

<!-- mawk '!/#/' results/step3ABQDPTHMQM.recode.vcf | wc -l -->

<!-- echo $? -->

<!-- # mapping quality -->
<!-- cut -f8 results/step3ABQDPTH.recode.vcf | grep -oe "MQM=[0-9]*" | sed -s 's/MQM=//g' > results/step3ABQDPTH.MQM -->
<!-- cut -f8 results/step3ABQDPTH.recode.vcf | grep -oe "MQMR=[0-9]*" | sed -s 's/MQMR=//g' > results/step3ABQDPTH.MQMR -->

<!-- ``` -->

# ```{r eval=FALSE}
# 
# temp <- read.table(here("filtering","results", "step3ABQDPTH.MQM"), col.names = "MQM")
# 
# mapqual <- read.table(here("filtering","results", "step3ABQDPTH.MQMR"), col.names = "MQMR")
# 
# mapqual <- bind_cols(mapqual, temp) %>%
#   mutate(ratio = MQM/MQMR)
# 
# filter <- mapqual %>%
#   filter(ratio < 0.25 | ratio > 1.75)
# 
# ggplot(mapqual, aes(x = MQM, y = MQMR)) +
#   geom_point(shape = 1) + 
#   geom_abline(intercept = 0, slope = 1, size = 1, color = "red", linetype = "dashed") +
#   geom_abline(intercept = 0, slope = 4, size = 1, color = "darkblue", linetype = "dashed") +
#   geom_abline(intercept = 0, slope = 0.571, size = 1, color = "darkblue", linetype = "dashed") +
#   geom_point(data = filter, aes(x = MQM, y = MQMR), shape = 1, color = "red") +
#   scale_x_continuous(limits = c(0, 65)) + 
#   scale_y_continuous(limits = c(0, 65)) +
#   labs(x = "mean mapping quality alt allele ", y = "mean mapping quality ref allele ") +
#   theme_standard
# 
# ```
It seems like this is cutting out a lot more loci than the red ones in the first graph.  I'm not sure the vcffilter is doing what it thinks it is.  Undo the cut of loci.



#### Maximum depth & Quality

Identify distribution of depth (based on original data set) to identify loci with excess coverage.

(INFO flags in filtered data set are are based on original number of individuals in data set).

Calculate average depth and standard deviation:

```{r plot depth vs qual}

# depth
depth <- read.table(here("filtering","results", "step3.DEPTH"),
                    col.names = "depth")

# quality score
qual <- read.table(here("filtering","results", "step3.QUAL"),
                   col.names = c("locus", "pos", "qual"))

# mean depth
mean_depth <- mean(depth$depth)

# standard deviation
std <- sd(depth$depth)

# mode
mode <- Mode(depth$depth)

cutoff <- sum(mean_depth + (2*std))

# identify SNP sites with depth > mean depth + 1 standard deviation and quality score < 2x the depth at that site
temp <- bind_cols(qual, depth) %>%
  filter(depth > cutoff) %>%
  filter(qual < 2*depth) %>%
  select(locus)

write.table(temp, here("filtering","results", "DEPTH.lowQloci"), col.names = FALSE, row.names = FALSE, quote = FALSE)

df <- bind_cols(qual, depth) %>%
  mutate(qualcutoff = 2*depth)

removeloc <- df %>%
  filter(depth > cutoff) %>%
  filter(qual < 2*depth)

ggplot(df, aes(x = depth, y = qual)) +
  geom_point(shape = 1) +
  geom_point(data = removeloc, aes(x = depth, y = qual), shape = 1, color = "red") +
  geom_line(data = df, aes(x = depth, y = qualcutoff), color = "blue",  linetype = "dashed", size = 1) +
  geom_vline(xintercept = cutoff, color = "blue", linetype = "dashed", size = 1) +
  labs(x = "") +
  theme_standard

```

<!-- Mean depth per locus (across all indivuals) is `r mean_depth` and the standard deviation is `r std`.  -->

<!-- Filter SNP site with depth > mean depth + 1 standard deviation = `r sum(mean_depth + 2*std)` and that have quality scores < 2x the depth at that site and output depth per site. -->

```{bash include=FALSE}

vcftools --vcf  results/step3ABQDPTH.recode.vcf --exclude-positions results/DEPTH.lowQloci --recode --recode-INFO-all --out results/step3ABQDPTHQUAL

vcftools --vcf results/step3ABQDPTHQUAL.recode.vcf --out results/step3ABQDPTHQUAL --site-mean-depth

```

Compare the distribution of mean depth per site averaged across individuals to determine cut-off value of sites with excessively high depth indicative of paralogs/multicopy loci.

```{r plot depth dist, message=FALSE, warning=FALSE}

# calculate mean depth per site (177 individuals)
depth <- read.table(here("filtering","results", "step3ABQDPTHQUAL.ldepth.mean"),header = TRUE, stringsAsFactors = FALSE)
  
# mean mean depth
mean <- mean(depth$MEAN_DEPTH)

# standard deviation
std <- sd(depth$MEAN_DEPTH)

# mode
mode <- Mode(depth$MEAN_DEPTH)

cutoff <- sum(mean + (2*std))

ggplot(depth, aes(x = MEAN_DEPTH)) +
  geom_histogram(binwidth = 5, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = mode),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = cutoff),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per site") +
  theme_standard

```

Choose cut-off for maximum mean read depth = 200 (calculate mean + 2x std: `r sum(mean + (2*std))`)

```{bash include=FALSE}

vcftools --vcf  results/step3ABQDPTHQUAL.recode.vcf --max-meanDP 200 --exclude-positions results/DEPTH.lowQloci --recode --recode-INFO-all --out results/FIL_6

# depth indv/locus
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --depth
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --site-mean-depth

# missing data indv/locus
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --missing-indv
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --missing-site

# heterozygosity per individual
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --het

# SNP call quality
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --site-quality

```

```{bash echo=FALSE}
  FILTER=$(echo "FIL_6_early")
  SNP=$(grep -cv '#' results/FIL_6.recode.vcf)
  CONTIG=$(grep -v '#' results/FIL_6.recode.vcf | cut -f 1 | sort | uniq | wc -l)
  INDV=$(vcfsamplenames results/FIL_6.recode.vcf | wc -l)
  echo "FIL_6.recode.vcf $FILTER $SNP $CONTIG $INDV" >> Filter.count


```

#### Plot stats after FIL_6 filters

```{r}

# load stats files ----
ind_stats_FIL_6 <- read.ind.stats(dir = "results", vcf = "FIL_6")

loc_stats_FIL_6 <- read.loc.stats(dir = "results", vcf = "FIL_6")

# plot missing data per indv ----
p1 <- ggplot(ind_stats_FIL_6, aes(x = F_MISS)) +
  geom_histogram(binwidth = .01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "missing data per indv") +
  theme_standard

# plot Fis per indv ----
p2 <- ggplot(ind_stats_FIL_6, aes(x = F)) +
  geom_histogram(binwidth = .01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(F, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "Fis per indv") +
  theme_standard

# plot read depth per indv ----
p3 <- ggplot(ind_stats_FIL_6, aes(x = MEAN_DEPTH)) +
  geom_histogram(binwidth = 10, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean read depth per indv") +
  theme_standard

# plot depth vs missing ----
p4 <- ggplot(ind_stats_FIL_6, aes(x = MEAN_DEPTH, y = F_MISS)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per indv", y = "% missing data") +
  theme_standard

# plot Fis vs missing data per indv ----
p5 <- ggplot(ind_stats_FIL_6, aes(x = F, y = F_MISS)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(F, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "Fis per indv", y = "% missing data") +
  theme_standard

# plot Fis vs mean depth per indv ----
p6 <- ggplot(ind_stats_FIL_6, aes(x = F, y = MEAN_DEPTH)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(F, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "Fis per indv", y = "mean depth per indv") +
  theme_standard

# plot distribution missing data per locus ----
p7 <- ggplot(loc_stats_FIL_6, aes(x = F_MISS)) +
  geom_histogram(binwidth = 0.01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0.1),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "% missing data per locus") +
  theme_standard

# plot distribution mean read depth ----
p8 <- ggplot(loc_stats_FIL_6, aes(x = MEAN_DEPTH)) +
  geom_histogram(binwidth = 5, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean read depth per locus") +
  theme_standard

# plot read depth vs missing data ----
p9 <- ggplot(loc_stats_FIL_6, aes(x = MEAN_DEPTH, y = F_MISS)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.1),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per locus", y = "% missing data") +
  theme_standard

# plot no of SNPs per locus ----
p10 <- loc_stats_FIL_6 %>%
  count(CHROM) %>%
  ggplot(aes(x = n)) +
  geom_histogram(binwidth = 1, color = "black", fill = "grey95") + 
  labs(x = "number of SNPs per locus") +
  theme_standard

temp <- loc_stats_FIL_6 %>%
  count(CHROM)

# plot number of SNPs per contig vs. mean depth ----
p11 <- left_join(temp, loc_stats_FIL_6) %>%
  ggplot() +
  geom_point(aes(x = n, y = MEAN_DEPTH)) +
  labs(x = "number of SNPs per contig", y = "mean depth") +
  theme_standard

# plot depth vs SNP quality ----
site_qual <- read.table(here("filtering","results", "FIL_6.lqual"), header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(PROB = 10^(-QUAL/10))

temp <- tibble(depth = loc_stats_FIL_6$MEAN_DEPTH, qual = site_qual$QUAL) 

p12 <- ggplot(temp, aes(x = depth, y = qual)) +
  geom_point(size = 1) +
  geom_vline(aes(xintercept = mean(depth, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(qual, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per locus", y = "SNP quality") +
  theme_standard

m6 <- multiplot(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, p11, p12, cols=2)

```

### Step 4: Filter Missing data loci/indv

Remove loci with genotype call rate < 60%.
```{bash include=FALSE}

vcftools --vcf results/FIL_6.recode.vcf --out results/step4a --max-missing 0.6 --recode --recode-INFO-all

vcftools --vcf results/step4a.recode.vcf --out results/step4a --missing-indv

```

```{bash echo=FALSE}
  FILTER=$(echo "step4a")
  SNP=$(grep -cv '#' results/step4a.recode.vcf)
  CONTIG=$(grep -v '#' results/step4a.recode.vcf | cut -f 1 | sort | uniq | wc -l)
  INDV=$(vcfsamplenames results/step4a.recode.vcf | wc -l)
  echo "step4a.recode.vcf $FILTER $SNP $CONTIG $INDV" >> Filter.count


```

Identify individuals with > 80% missing data

```{r}

imiss <- read.table(here("filtering","results", "step4a.imiss"), 
                    header = TRUE, stringsAsFactors = FALSE)

ggplot(imiss, aes(x = F_MISS)) +
  geom_histogram(binwidth = 0.05, color = "black", fill = "grey75") +
  geom_vline(xintercept = .8, color = "darkred", linetype = "dashed") +
  scale_x_continuous(limits = c(0, 1)) +
  theme_standard

LQ_indv <- imiss %>%
  filter(F_MISS > 0.8) %>%
  select(INDV)

write.table(LQ_indv, here("filtering","results", "LQ_Ind_FIL_6a"),
            col.names = FALSE, row.names = FALSE, quote = FALSE)

```

Remove flagged individuals

```{bash include=FALSE}

vcftools --vcf results/step4a.recode.vcf --out results/step4b --remove results/LQ_Ind_FIL_6a --recode --recode-INFO-all

```

```{bash echo=FALSE}
  FILTER=$(echo "step4b")   
  SNP=$(grep -cv '#' results/step4b.recode.vcf)
  CONTIG=$(grep -v '#' results/step4b.recode.vcf | cut -f 1 | sort | uniq | wc -l)
  INDV=$(vcfsamplenames results/step4b.recode.vcf | wc -l)
  echo "step4b.recode.vcf $FILTER $SNP $CONTIG $INDV" >> Filter.count
```

## Step 5 
Remove loci with genotype call rate < 70%.
```{bash include=FALSE}
vcftools --vcf results/step4b.recode.vcf --out results/step5a --max-missing 0.7 --recode --recode-INFO-all

vcftools --vcf results/step5a.recode.vcf --out results/step5a --missing-indv

```

```{bash echo=FALSE}
  FILTER=$(echo "step5a")   
  SNP=$(grep -cv '#' results/step5a.recode.vcf)
  CONTIG=$(grep -v '#' results/step5a.recode.vcf | cut -f 1 | sort | uniq | wc -l)
  INDV=$(vcfsamplenames results/step5a.recode.vcf | wc -l)
  echo "step5a.recode.vcf $FILTER $SNP $CONTIG $INDV" >> Filter.count
```

Identify individuals with > 70% missing data

```{r}

imiss <- read.table(here("filtering","results", "step5a.imiss"), 
                    header = TRUE, stringsAsFactors = FALSE)

ggplot(imiss, aes(x = F_MISS)) +
  geom_histogram(binwidth = 0.05, color = "black", fill = "grey75") +
  geom_vline(xintercept = .7, color = "darkred", linetype = "dashed") +
  scale_x_continuous(limits = c(0, 1)) +
  theme_standard

LQ_indv <- imiss %>%
  filter(F_MISS > 0.7) %>%
  select(INDV)

write.table(LQ_indv, here("filtering","results", "LQ_Ind_FIL_6b"),
            col.names = FALSE, row.names = FALSE, quote = FALSE)

```

Remove flagged individuals

```{bash include=FALSE}

vcftools --vcf results/step5a.recode.vcf --out results/step5b --remove results/LQ_Ind_FIL_6b --recode --recode-INFO-all

```

```{bash echo=FALSE}


  FILTER=$(echo "step5b")   
  SNP=$(grep -cv '#' results/step5b.recode.vcf)
  CONTIG=$(grep -v '#' results/step5b.recode.vcf | cut -f 1 | sort | uniq | wc -l)
  INDV=$(vcfsamplenames results/step5b.recode.vcf | wc -l)
  echo "step5b.recode.vcf $FILTER $SNP $CONTIG $INDV" >> Filter.count


```



Final FIL_6 filtering step, remove indels and apply maf of 0.2
```{bash include=FALSE}

vcfallelicprimitives results/step5b.recode.vcf --keep-info --keep-geno > results/SNPs.vcf

vcftools --vcf results/SNPs.vcf --out results/FIL_6 --remove-indels --maf 0.2 --recode --recode-INFO-all

# depth indv/locus
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --depth
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --site-mean-depth

# missing data indv/locus
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --missing-indv
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --missing-site

# heterozygosity per individual
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --het

# SNP call quality
vcftools --vcf results/FIL_6.recode.vcf --out results/FIL_6 --site-quality

```

### Plot stats for filtered data set

```{r}

# load stats files ----
ind_stats_FIL_6 <- read.ind.stats(dir = "results", vcf = "FIL_6")

loc_stats_FIL_6 <- read.loc.stats(dir = "results", vcf = "FIL_6")

# plot missing data per indv ----
p1 <- ggplot(ind_stats_FIL_6, aes(x = F_MISS)) +
  geom_histogram(binwidth = .01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "missing data per indv") +
  theme_standard

# plot Fis per indv ----
p2 <- ggplot(ind_stats_FIL_6, aes(x = F)) +
  geom_histogram(binwidth = .01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(F, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "Fis per indv") +
  theme_standard

# plot read depth per indv ----
p3 <- ggplot(ind_stats_FIL_6, aes(x = MEAN_DEPTH)) +
  geom_histogram(binwidth = 10, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean read depth per indv") +
  theme_standard

# plot depth vs missing ----
p4 <- ggplot(ind_stats_FIL_6, aes(x = MEAN_DEPTH, y = F_MISS)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per indv", y = "% missing data") +
  theme_standard

# plot Fis vs missing data per indv ----
p5 <- ggplot(ind_stats_FIL_6, aes(x = F, y = F_MISS)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(F, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "Fis per indv", y = "% missing data") +
  theme_standard

# plot Fis vs mean depth per indv ----
p6 <- ggplot(ind_stats_FIL_6, aes(x = F, y = MEAN_DEPTH)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(F, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "Fis per indv", y = "mean depth per indv") +
  theme_standard

# plot distribution missing data per locus ----
p7 <- ggplot(loc_stats_FIL_6, aes(x = F_MISS)) +
  geom_histogram(binwidth = 0.01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0.1),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "% missing data per locus") +
  theme_standard

# plot distribution mean read depth ----
p8 <- ggplot(loc_stats_FIL_6, aes(x = MEAN_DEPTH)) +
  geom_histogram(binwidth = 5, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean read depth per locus") +
  theme_standard

# plot read depth vs missing data ----
p9 <- ggplot(loc_stats_FIL_6, aes(x = MEAN_DEPTH, y = F_MISS)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(MEAN_DEPTH, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.1),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per locus", y = "% missing data") +
  theme_standard

# plot no of SNPs per locus ----
p10 <- loc_stats_FIL_6 %>%
  count(CHROM) %>%
  ggplot(aes(x = n)) +
  geom_histogram(binwidth = 1, color = "black", fill = "grey95") + 
  labs(x = "number of SNPs per locus") +
  theme_standard

temp <- loc_stats_FIL_6 %>%
  count(CHROM)

# plot number of SNPs per contig vs. mean depth ----
p11 <- left_join(temp, loc_stats_FIL_6) %>%
  ggplot() +
  geom_point(aes(x = n, y = MEAN_DEPTH)) +
  labs(x = "number of SNPs per contig", y = "mean depth") +
  theme_standard

# plot depth vs SNP quality ----
site_qual <- read.table(here("filtering","results", "FIL_6.lqual"), header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(PROB = 10^(-QUAL/10))

temp <- tibble(depth = loc_stats_FIL_6$MEAN_DEPTH, qual = site_qual$QUAL) 

p12 <- ggplot(temp, aes(x = depth, y = qual)) +
  geom_point(size = 1) +
  geom_vline(aes(xintercept = mean(depth, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(qual, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per locus", y = "SNP quality") +
  theme_standard

m7 <- multiplot(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, p11, p12, cols=2)

```

Data set contains `r nrow(ind_stats_FIL_6)` individuals and `r nrow(loc_stats_FIL_6)` loci.


## Run SNP counting script
Compare SNPs/contigs/indv at each filtering step and between filtering schemes.

```{bash include=FALSE}


  FILTER=$(echo "FIL_6_later")   
  SNP=$(grep -cv '#' results/FIL_6.recode.vcf)
  CONTIG=$(grep -v '#' results/FIL_6.recode.vcf | cut -f 1 | sort | uniq | wc -l)
  INDV=$(vcfsamplenames results/FIL_6.recode.vcf | wc -l)
  echo "FIL_6.recode.vcf $FILTER $SNP $CONTIG $INDV" >> Filter.count
```

```{r}
count <- read_table(here("filtering","Filter.count"))
knitr::kable(count)
```
Convert the vcf to a genepop
```{bash eval=FALSE, include=FALSE}
vcf2genepop.pl vcf=results/FIL_6.recode.vcf > seq33_03_baits_only_SNPs.gen
```



